source/common/http/conn_manager_impl.cc

```diff
@@ -621,17 +621,14 @@ ConnectionManagerImpl::ActiveStream::~ActiveStream() {
     }
   }
 
-  // TODO(alyssawilk) this is not true. Fix.
-  // A downstream disconnect can be identified for HTTP requests when the upstream returns with a 0
-  // response code and when no other response flags are set.
-  if (!filter_manager_.streamInfo().hasAnyResponseFlag() &&
-      !filter_manager_.streamInfo().responseCode()) {
-    filter_manager_.streamInfo().setResponseFlag(
-        StreamInfo::ResponseFlag::DownstreamConnectionTermination);
-  }
   if (connection_manager_.remote_close_) {
     filter_manager_.streamInfo().setResponseCodeDetails(
         StreamInfo::ResponseCodeDetails::get().DownstreamRemoteDisconnect);
+    if (!filter_manager_.streamInfo().hasAnyResponseFlag() &&
+        !filter_manager_.streamInfo().responseCode()) {
+      filter_manager_.streamInfo().setResponseFlag(
+          StreamInfo::ResponseFlag::DownstreamConnectionTermination);
+    }
```

Comment:
Admittedly I am the cause of this confusion from years ago, but I'm having a hard time understanding the current status quo wrt these flags. Can you potentially add more comments about what all of this means and what is still broken so we can track? I'm not exactly sure what "downstream connection termination" means in this context at this point. Are there any other docs that need to be updated also related to this (access logging?)?

Reply:
Yeah, I was inclined to file an issue, but I wasn't really sure what the goal was, basically if it was meant to be opt-in, or if it was meant to be more like details and set everywhere?  Let's sort out a plan and I can file a tech debt issue.

The replaced code was supposed to be a catchall for no RC set, but  as commented, there's a bunch of ways Envoy can tear down streams with no RC which are not downstream disconnect - downstream H2 reset frames, max duration timeout etc.

Even if we fixed that,  it's really not guaranteed that any flag is set on sendLocalReply, or encodeHeaders (another filter like routing or caching filter) so there's loads of ways to have a stream without flags set.  I didn't know if this was meant to be an opt-in sort of thing, where filters (like rate limit) *should* add them, or if we would just want a catchall for local replies?.  Based on details I think there's probably O(100) ways Envoy can send responses or resets, and about 20 flags (though many of those are catchalls for several details, like the details specify codec error where the flags just indicate codec error) so if I had to guess I'd say 30% of the paths with no flags?  Those should all have details though, so ideally folks can still debug.  If there's value in catch-all "this was a stream reset"/"this was a disconnect"/"this was sendLocalReply" we could do that (though it's a bit weird that some disconnects like idle timeouts get their own flags and others don't).  Alternately we could add response flags alongside details to all the reset/disconnect/localreply APIs to require flags be set for all of those calls without catchalls, but that's a bunch more work.

Reply:
My personal feeling would be to just delete this, given that we set DownstreamRemoteDisconnect right above this, which I think was the original intention of this sequence of code, but honestly I don't fully recall the history. I don't think this flag adds much value as you point out, and then over time we an add more specific flags as needed? Is the issue that if we delete this we hit the other ASSERT you added? Should we somehow make the assert work with no flag set here?

Reply:
I think we don't want to lose (accurate) information in debug flags given that a lot of people log flags but not details.  I think fixing it makes more sense than removing it?

Reply:
Ohh. Sorry, I was confusing setResponseCodeDetails() with setResponseFlag(). :facepalm: That was a lot of my confusion.

OK this makes sense for now. Can you add more comments just about this flow and what you recommend moving forward?

Reply:
I don't think the fix needs comments - I think it's a fine standalone fix.

I think if we want a tracking issue for cleaning up response flags for consistent setting the way I have been with details we can, or we can just add flags as folks find them useful. :-)

Reply:
Sorry what I'm asking for is just comments on this code block. Can you add a comment on why we only set this flag if there is no other flag set and response code is zero? I think that would aid the next reader.

Reply:
Oh, sure, that makes sense.
Actually I have a TODO for you somewhere asking if it's Ok to set multiple flags.  IMO if Envoy sends a cached response and we get a downstream disconnect, it'd be better to have both flags set, so I'd be inclined to just set this unilaterally if there's a downstream remote disconnect.  Originally I was just going for the minimally correct change, but I think 2 flags indicating what actually happened is less confusing than one flag only set when there's no other flag, WDYT?

Reply:
Yes +1. I think that would make a lot more sense. That might need a release note but I would be fine without feature flagging it.